sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -67,7 +67,7 @@ trait OperationHelper extends AliasHelper with PredicateHelper {
           empty
         }
 
-      case Filter(condition, child) =>
+      case Filter(condition, child) if !condition.exists(_.isInstanceOf[PythonUDF]) =>
```

Comment:
shall we put the fix in `PruneFileSourcePartitions` instead of this low-level pattern match? For example, we exclude subquery expressions in `PruneFileSourcePartitions`

Reply:
The benefit is we can still prune partitions with `WHERE udf(c1) = 1 AND c2 = 2`

Reply:
SGTM but that would be more and less same as https://github.com/apache/spark/pull/47033. Let me take another look